lib/Dialect/AIE/IR/AIEDialect.cpp

```diff
@@ -844,28 +838,59 @@ LogicalResult xilinx::AIE::MemOp::verify() {
 xilinx::AIE::TileOp xilinx::AIE::MemOp::getTileOp() {
   return cast<xilinx::AIE::TileOp>(getTile().getDefiningOp());
 }
-int xilinx::AIE::MemOp::colIndex() {
-  Operation *Op = getTile().getDefiningOp();
-  xilinx::AIE::TileOp tile = dyn_cast<xilinx::AIE::TileOp>(Op);
+int xilinx::AIE::MemOp::colIndex() { return getTileOp().colIndex(); }
 
-  return tile.colIndex();
-}
+int xilinx::AIE::MemOp::rowIndex() { return getTileOp().rowIndex(); }
 
-int xilinx::AIE::MemOp::rowIndex() {
-  Operation *Op = getTile().getDefiningOp();
-  xilinx::AIE::TileOp tile = dyn_cast<xilinx::AIE::TileOp>(Op);
+/// Returns the region on the current operation that is callable. This may
+/// return null in the case of an external callable object, e.g. an external
+/// function.
+Region *xilinx::AIE::MemOp::getCallableRegion() { return &(getBody()); }
 
-  return tile.rowIndex();
+/// Returns the results types that the callable region produces when executed.
+ArrayRef<Type> xilinx::AIE::MemOp::getCallableResults() { return getType(); }
+
+// MemTileDMAOp
+LogicalResult xilinx::AIE::MemTileDMAOp::verify() {
+  assert(getOperation()->getNumRegions() == 1 &&
+         "MemTileDMAOp has zero region!");
+  assert(!getBody().empty() && "MemTileDMAOp should have non-empty body");
+
+  auto result =
+      HasSomeTerminator<xilinx::AIE::DMAStartOp, xilinx::AIE::NextBDOp,
+                        xilinx::AIE::EndOp>::verifyTrait(*this);
+  if (result.failed()) {
+    return result;
+  }
+
+  for (auto &bodyOp : getBody().getOps()) {
+    if (auto allocOp = dyn_cast<memref::AllocOp>(bodyOp)) {
+      if (!allocOp->getAttr("id"))
+        return allocOp.emitOpError()
+               << "allocOp in MemTileDMAOp region should have an id attribute";
+    }
+  }
+  return success();
+}
+
+xilinx::AIE::TileOp xilinx::AIE::MemTileDMAOp::getTileOp() {
+  return cast<xilinx::AIE::TileOp>(getTile().getDefiningOp());
 }
+int xilinx::AIE::MemTileDMAOp::colIndex() { return getTileOp().colIndex(); }
+
+int xilinx::AIE::MemTileDMAOp::rowIndex() { return getTileOp().rowIndex(); }
 
 /// Returns the region on the current operation that is callable. This may
 /// return null in the case of an external callable object, e.g. an external
```

Comment:
```suggestion
/// return nullptr in the case of an external callable object, e.g. an external
```
Do you think it is possible to have better code factorization in the class hierarchy (mixin?) to remove code repetition. Just some random proposition since I do not understand the big picture...

Reply:
perhaps... patches welcome?